rfcs/0007-pass-spec.md

```diff
@@ -0,0 +1,453 @@
+
+For most passes, we should switch from requiring that widgets recurse the same methods to all their children, to having `WidgetPod` directly call those methods on the children.
+
+This will mostly involve changes to `WidgetPod`'s internals.
+
+For targeted passes, we should add a private method to RenderRoot which directly sends an arbitrary event to a specific widget, then propagates changes upwards.
+
+### Add MUTATE pass
+
+We should add a queue of callbacks with their target widget.
+Calling `FoobarCtx::mutate_later()` would add a callback to that queue.
+
+The MUTATE pass would then go through that queue.
+
+### Add compose pass
+
+We should add a `compose()` method to the WidgetTrait, a `WidgetPod::compose` method, a `ComposeCtx` type, a `RenderRoot::root_compose()` method, etc.
+
+The `ParentWindowOrigin` lifecycle event should be replaced by this compose pass.
+
+Context types should get a `request_compose()` method.
+
+### Remove LifecycleCtx methods
+
+The following methods specific to LifecycleCtx should be removed, and replaced with getter methods on the Widget trait:
+
+- `register_for_focus`
+- `register_as_text_input`
+- `register_as_portal`
+
+For instance, when a widget is created, `WidgetPod` should call a method called `Widget::accepts_focus` on the widget.
+If so, it should add it to the set of widgets accepting focus.
+
+Then when the widget is deleted, it should be removed from that set.
+
+### Send lifecycle events directly from RenderRoot
+
+The following passes should be created in RenderRoot, which will send LifecycleEvents directly to the concerned widgets:
+
+- `UPDATE_TREE`: Updates internal flags and sends RequestPanToChild.
+- `UPDATE_FOCUS`: Sends FocusChanged event.
+- `UPDATE_DISABLED`: Sends DisabledChanged event.
+- `UPDATE_ANIM`: Sends AnimFrame event.
+- `UPDATE_SCROLLS`: Sends RequestPanToChild.
+- `UPDATE_POINTER`: Sends HotChanged and updates the cursor icon.
+
+These features should progressively be moved out of the current `WidgetPod::lifecycle` method.
+
+### Replace Lifecycle and StatusChange with StatusUpdate event
+
+The previous sections are about progressive changes that should be implemented to take things off the lifecycle pass.
+
+The lifecycle pass and the Lifecycle event should eventually be replaced with the UPDATE passes and StatusUpdate event altogether.
+
+### Rework paint pass
+
+The paint pass should now build the root scene in a single pass.
+
+`WidgetPod::paint()` will do the following:
+
+- Call the inner paint().
+- That paint "returns" both a scene fragment and an optional clip layer.
+- Append the scene fragment to the global scene.
+- If the clip layer is Some, push it.
+- For each child, call `WidgetPod::paint()` with the transform that was set in `compose()`.
+- If the clip layer is Some, pop it.
+
+Unlike the current algorithm where fragments are added to parent fragments, here they are added directly to the root.
+This is `O(N)` instead of `O(N * Depth)`.
```

Comment:
I know what you're talking about here, but I don't think a casual reader would understand this comment.

Maybe linking to [#xilem > The mystery of the Xilem bomb](https://xi.zulipchat.com/#narrow/stream/354396-xilem/topic/The.20mystery.20of.20the.20Xilem.20bomb) is sufficient?